hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsUriQueryBuilder.java

```diff
@@ -59,6 +64,20 @@ public String toString() {
         throw new IllegalArgumentException("Query string param is not encode-able: " + entry.getKey() + "=" + entry.getValue());
       }
     }
+    // append SAS Token
+    if (sasToken != null) {
+      sasToken = sasToken.startsWith(AbfsHttpConstants.QUESTION_MARK)
+          ? sasToken.substring(1)
+          : sasToken;
```

Comment:
The interface should require that '?' be omitted from the SAS token.  Allowing '?' means the plug-in will add it and then driver will remove it every time, which is wasteful.

Reply:
Uribuilder toString() will auto add '?' at the start. Like in the case of the test SASGenerator code added to this PR, the SAS token generator would end up having a '?'. SAS token being a query URL it might be better to handle to presence or absence for '?' in the received token ?

Reply:
Removed the handling of starting '?' in SAS Token query.